packages/server/src/orchestrator/state.ts

```diff
@@ -50,26 +49,19 @@ export type BundlerState =
 
 export type BundlerTypes = Pick<BundlerState, 'type'>['type'];
 
-export type ServiceStatus =
-  | { type: 'unstarted' }
-  | { type: 'started' }
-  | { type: 'reachable' }
-  | { type: 'exited'; exitStatus: ExitStatus };
-
-export type ServiceStatuses = ServiceStatus['type'];
-
-export type ServiceState<O> = {
-  id: string;
-  name: string;
-  status: ServiceStatus;
-  // TODO: we need to further constrain this to Slice(s)
-  atom?: Atom<OrchestratorState>;
-} & O;
+export type ServiceStatus = {
+  type: string;
+};
 
-export type Service<O> = {
-  (options: Partial<Omit<ServiceState<O>, keyof O>> & O): Operation<void>;
+export type ServiceState<S extends ServiceStatus, O> = {
+  options: O;
+  status: S;
 };
```

Comment:
When in comes to the Service State I think that the guiding principle should be that the intrinsic properties of the state should be those that are managed by the orchestrator, and those portions of the state that vary are the ones about which only the particular service has knowledge. If we use that as a heuristic, then what does the orchestrator know?

It knows 1) when it has not even started the service operation as in it has not yet invoked it. It knows 2) it has invoked the service operation and 3a) when the service operation has completed or 3b) when the service operation has raised an error. This would seem to imply that the intrinsitc statuses are (for example):

`unstarted`, `started`, `completed`, `errored`

> Worth noting that `completed` is a useful piece of information since I've definitely been bitten by a service operation running to completion too quickly because of spawning a bunch of things and then not having an actual blocking operation.

Everything else can only be known internally as the service operation runs and so falls under the purview of the service itself, and so it's _that_ state that should form the basis for `S`. It does seem odd not to have a `ready` state in the above list, but readiness is something that can only be known by the service. We could make an exception by having the orchestrator pass a ready callback so that while technically it is managing that property, it is only via inversion of control.

Then, the service actually does not receive the service state, but only a slice of `S` and everything else would be managed by the orchestrator. Under that regimen, `S` would be completely unconstrained. 

I'd propose that the services are invoked with their options,  S is unconstrained, and that services are run in a uniform way by the orchestrator with a `manageService` operation (which will eventually restart the service when the options slice changes). Something like:

```ts
export interface ServiceState<S, O> {
  status = 'unstarted' | 'started' | 'completed' | 'errored';
  state: S;
  options: 0;
  error?: unknown;
}
```

That's the endgame that I think we're shooting for (with option of having a `ready()` callback). I think that we've got mostly that here, so I would just rename `ServiceStatus` to something like `props`, `attrs`, or `state` to indicate that it is completely owned by the service.

Reply:
@jnicklas @cowboyd I think this is slightly going back where I was 2 commits ago :).

I think we need to somehow reach an agreement with what the right thing is without writing code.  Not sure how we do that.

a small minor note, is that `Error` should be a generic type argument.  Then that covers the face of `ExitStatus` or `xhrError` or whatever.

```ts
export interface ServiceState<S, O, E = Error> {
  status = 'unstarted' | 'started' | 'completed' | 'errored';
  state: S;
  options: 0;
  error?: E;
}
```